ledger/src/rand_compat.rs

```diff
@@ -0,0 +1,248 @@
+        for weight in weights.iter_mut() {
+            total_weight += *weight;
+            *weight = total_weight;
+        }
+        if weights.pop().is_none() {
+            return Err(WeightedError::NoItem);
+        }
+        if total_weight == 0 {
+            return Err(WeightedError::AllWeightsZero);
+        }
+
+        Ok(Self {
+            weights,
+            total_weight_sampler: UniformU64Sampler::new_like_instance_sample(total_weight),
+        })
+    }
+
+    pub fn sample(&self, rng: &mut impl Rng) -> usize {
+        let chosen_weight = self.total_weight_sampler.sample(rng);
+        // Find the first item which has a weight *higher* than the chosen weight.
+        self.weights
+            .binary_search_by(|w| {
+                if *w <= chosen_weight {
+                    Ordering::Less
+                } else {
+                    Ordering::Greater
+                }
+            })
+            .unwrap_err()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        super::*,
+        assert_matches::assert_matches,
+        rand0_8_5::{
+            distributions::{
+                uniform::{SampleUniform, UniformSampler as _},
+                WeightedIndex,
+            },
+            prelude::Distribution,
+            SeedableRng as _,
+        },
+        rand_chacha0_3_1::ChaChaRng,
+        sha2::{Digest, Sha256},
+        std::array,
+        test_case::test_case,
+    };
+
+    const CHACHA_SEED: [u8; 32] = [16; 32];
+
+    #[test]
+    fn test_uniform_sample_like_instance_sample_example() {
+        let mut rng_compat = ChaChaRng::from_seed(CHACHA_SEED);
+        let sampler_compat = UniformU64Sampler::new_like_instance_sample(294_533);
+        let values: [u64; 10] = array::from_fn(|_| sampler_compat.sample(&mut rng_compat));
+        assert_eq!(
+            values,
+            [280405, 7507, 84194, 272634, 52124, 190984, 8676, 230277, 223574, 126007]
+        );
+    }
+
+    #[test_case(10, "5p3DrG89DLrJotbMuZJUDHMrqcfoQiDQpa3FbNDbnND6")]
+    #[test_case(2_729, "J65XfpzkppuxWEyvgVdDmGBsJSex6BuevYuDSRtaDNAK")]
+    #[test_case(4_098, "EaRABrcPBw5LLNBmjn5ay5YN5yTPgv3GnQJLbpUarRkJ")]
+    #[test_case(504_302_479, "HniJPVe7zir8XxHmtUuxzhPVvWjMcJxVhhj8QSs1PZTC")]
+    #[test_case(1_000_346_000_000, "BghMy7yLe6BVzMbc4zNvAB4sz3ZSPYKJS5oLGvXYVzw2")]
+    fn test_uniform_sampler_like_instance_sample_compat(range_end: u64, expected_hash: &str) {
+        let mut rng_rand = ChaChaRng::from_seed(CHACHA_SEED);
+        let sampler_rand = <u64 as SampleUniform>::Sampler::new(0, range_end);
+
+        let mut rng_compat = ChaChaRng::from_seed(CHACHA_SEED);
+        let sampler_compat = UniformU64Sampler::new_like_instance_sample(range_end);
+
+        let mut hash = Sha256::new();
+        (0..600_000).for_each(|i| {
+            let rand = sampler_rand.sample(&mut rng_rand);
+            let compat = sampler_compat.sample(&mut rng_compat);
+            assert_eq!(rand, compat, "should be equal at {i}");
```

Comment:
This assert will go away once we switch to new `rand` version, since it will not produce the same errors. The seeds and number of iterations is actually picked to trigger the mismatched value between rand versions.
Because of above, the test includes explicit hash of values and checks for that - this should remain stable after we move to new rand version.